File: web/src/api/alert.ts
import { http } from './request'

export type AlertSeverity = 'p0' | 'p1' | 'p2' | 'p3' | 'info'
export type AlertStatus =
  | 'new'
  | 'acknowledged'
  | 'processing'
  | 'recovered'
  | 'closed'
  | 'silenced'

export type AlertEventType =
  | 'triggered'
  | 'updated'
  | 'recovered'
  | 'acknowledged'
  | 'assigned'
  | 'processing_started'
  | 'closed'
  | 'silenced'
  | 'unsilenced'
  | 'commented'
  | 'ai_analysis_requested'
  | 'execution_created'
  | 'execution_started'
  | 'execution_finished'

export interface Alert {
  id: string
  external_id?: string
  source: string
  source_id?: string
  source_name?: string
  fingerprint: string
  dedup_key: string
  name: string
  summary?: string
  description?: string
  severity: AlertSeverity | string
  status: AlertStatus | string
  rule_id?: string
  rule_name?: string
  business_line?: string
  environment?: string
  application_id?: string
  application_name?: string
  resource_id?: string
  resource_type?: string
  resource_name?: string
  owner_user_id?: string
  assignee_user_id?: string
  labels: Record<string, string>
  annotations: Record<string, string>
  occurrence_count: number
  first_seen_at: number
  last_seen_at: number
  recovered_at?: number
  acknowledged_at?: number
  closed_at?: number
  silenced_until?: number
  created_at: number
  updated_at: number
}

export interface AlertEvent {
  id: string
  alert_id: string
  event_type: AlertEventType | string
  actor_type: string
  actor_id?: string
  actor_name?: string
  message?: string
  payload: Record<string, unknown>
  created_at: number
}

export interface AlertDetail {
  alert: Alert
  events: AlertEvent[]
  related: Record<string, unknown>
}

export interface PageResult<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

export interface AlertListQuery {
  page?: number
  page_size?: number
  status?: string
  severity?: string
  source?: string
  source_id?: string
  business_line?: string
  environment?: string
  application_id?: string
  resource_id?: string
  assignee_user_id?: string
  keyword?: string
  active_only?: boolean
  from?: number
  to?: number
}

export interface AlertSource {
  id: string
  name: string
  type: string
  enabled: boolean
  secret_masked?: string
  environment?: string
  business_line?: string
  description?: string
  created_at: number
  updated_at: number
}

export interface CreateAlertSourceInput {
  id: string
  name: string
  type?: string
  enabled?: boolean
  secret: string
  environment?: string
  business_line?: string
  description?: string
}

export interface UpdateAlertSourceInput {
  name?: string
  type?: string
  enabled?: boolean
  secret?: string
  environment?: string
  business_line?: string
  description?: string
}

export interface AIAnalysisInput {
  time_range?: string
  include_logs?: boolean
  include_metrics?: boolean
  include_changes?: boolean
}

export function listAlerts(params?: AlertListQuery) {
  return http<PageResult<Alert>>({
    url: '/api/alerts',
    method: 'get',
    params
  })
}

export function getAlert(alertId: string) {
  return http<AlertDetail>({
    url: `/api/alerts/${encodeURIComponent(alertId)}`,
    method: 'get'
  })
}

export function acknowledgeAlert(alertId: string, message?: string) {
  return http<Alert>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/acknowledge`,
    method: 'post',
    data: message ? { message } : {}
  })
}

export function assignAlert(alertId: string, assigneeUserId: string, message?: string) {
  return http<Alert>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/assign`,
    method: 'post',
    data: { assignee_user_id: assigneeUserId, message }
  })
}

export function startProcessingAlert(alertId: string) {
  return http<Alert>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/start-processing`,
    method: 'post',
    data: {}
  })
}

export function recoverAlert(alertId: string, message?: string) {
  return http<Alert>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/recover`,
    method: 'post',
    data: message ? { message } : {}
  })
}

export function closeAlert(alertId: string, resolution: string) {
  return http<Alert>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/close`,
    method: 'post',
    data: { resolution }
  })
}

export function silenceAlert(alertId: string, reason: string, durationS: number) {
  return http<Alert>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/silence`,
    method: 'post',
    data: { reason, duration_s: durationS }
  })
}

export function unsilenceAlert(alertId: string) {
  return http<Alert>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/unsilence`,
    method: 'post',
    data: {}
  })
}

export function commentAlert(alertId: string, message: string) {
  return http<AlertEvent>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/comments`,
    method: 'post',
    data: { message }
  })
}

export function requestAlertAIAnalysis(alertId: string, input?: AIAnalysisInput) {
  return http<AlertEvent>({
    url: `/api/alerts/${encodeURIComponent(alertId)}/ai-analysis`,
    method: 'post',
    data: input ?? {}
  })
}

export function listAlertSources() {
  return http<{ items: AlertSource[] }>({
    url: '/api/alerts/sources',
    method: 'get'
  })
}

export function getAlertSource(sourceId: string) {
  return http<AlertSource>({
    url: `/api/alerts/sources/${encodeURIComponent(sourceId)}`,
    method: 'get'
  })
}

export function createAlertSource(input: CreateAlertSourceInput) {
  return http<AlertSource>({
    url: '/api/alerts/sources',
    method: 'post',
    data: input
  })
}

export function updateAlertSource(sourceId: string, input: UpdateAlertSourceInput) {
  return http<AlertSource>({
    url: `/api/alerts/sources/${encodeURIComponent(sourceId)}`,
    method: 'put',
    data: input
  })
}

export function deleteAlertSource(sourceId: string) {
  return http<{ deleted: boolean }>({
    url: `/api/alerts/sources/${encodeURIComponent(sourceId)}`,
    method: 'delete'
  })
}

File: internal/alert/domain/alert.go
package domain

import "time"

// Alert 是告警中心主记录，一条记录代表一轮可处理的告警生命周期。
// 字段与 alert_alert 表及 API Alert 对象对齐；application_id/resource_id 为 AI/Execution 预留关联。
type Alert struct {
	ID              string            // 告警业务 ID（UUID）
	ExternalID      string            // 外部告警 ID 或 fingerprint
	Source          string            // 来源类型，如 prometheus_alertmanager
	SourceID        string            // 平台接入源 ID
	SourceName      string            // 接入源显示名快照
	Fingerprint     string            // 外部指纹
	DedupKey        string            // 平台去重键，必须稳定
	LifecycleSeq    int               // 生命周期序号，closed 后重开递增
	Name            string            // 告警名称
	Summary         string            // 摘要
	Description     string            // 详细描述
	Severity        AlertSeverity     // 级别
	Status          AlertStatus       // 当前状态
	RuleID          string            // 外部规则 ID
	RuleName        string            // 外部规则名
	BusinessLine    string            // 业务线
	Environment     string            // 环境
	ApplicationID   string            // 关联应用 ID（Asset 匹配后写入）
	ApplicationName string            // 关联应用名快照
	ResourceID      string            // 关联资源 ID
	ResourceType    string            // 资源类型
	ResourceName    string            // 资源名称
	OwnerUserID     string            // 归属负责人
	AssigneeUserID  string            // 当前处理人
	Labels          map[string]string // 标准化标签
	Annotations     map[string]string // 标准化注解
	OccurrenceCount int               // 重复触发次数
	FirstSeenAt     time.Time         // 首次触发
	LastSeenAt      time.Time         // 最近触发/更新
	RecoveredAt     *time.Time        // 恢复时间
	AcknowledgedAt  *time.Time        // 认领时间
	ClosedAt        *time.Time        // 关闭时间
	SilencedUntil   *time.Time        // 静默截止
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

// IsActive 返回告警是否仍处于未关闭状态（含 silenced，仍接收外部更新）。
func (a Alert) IsActive() bool {
	return a.Status.IsActive()
}
